### ISSUE TITLE
WebSocket Server - Intercept payload send/recieve

### ISSUE THREAD

**Issue by goddanao** [swoft-cloud/swoft#862]
**Feature request**

Trying to implement the "permessage-deflate" WebSocket extension i need a way to process in/out frames to apply appropriate compression/decompression. 

There's no WsServerEvent like MESSAGE_SEND and MESSAGE_RECIEVE to intercept and process frames (in the default wsserver implementation the frames are immedialy sent by the internal swoole sever).

There's a chance to add this feature or, maybe in the future, you'll add some sort of websocket extensions support that covers these scenarios?

Thankyou!

**Comment by inhere** [comment]
hi @goddanao 

Thank you very much for your suggestion, we will be adding the relevant `event` support.

Please pay attention to the release of `2.0.5`.

**Comment by goddanao** [comment]
hi @inhere ,
happy to help as i can. In the event support you're going to add, please make ws opcode changeable, since the compressed stream needs to be binary.

I'm impatient for your next release.
Thank you

**Comment by inhere** [comment]
hi @goddanao   

The `v2.0.5` has been released. please try it. Hope that has solved your problem.

**Comment by goddanao** [comment]
hi @inhere, 
thankyou i see that you add 'swoft.ws.server.message.send' with $response as target, which it's what i need to modify before sending data over the wire. 

I still need a way to intercept incoming packets, some sort of  'swoft.ws.server.message.recieve' for decoding packets at connection level (since protocol/extensions are negotiated per connection), before passing them to WsModule / Parsers.

I think it could be useful for easy implementation on ws protocol/extension.

**Comment by goddanao** [comment]
hi again @inhere,
i think the event WsServerEvent::MESSAGE_SEND won't fit for my idea of intercepting Frames coming and before going over the wire. 
Is called too early and it's too "high level", any Message->data manipulation will be later "formatted" in  the Response before sending it, causing the Parser to elaborate, in my case, an inflated binary stream and crashing. 

Maybe is better introduce WsServerEvent::FRAME_SEND and WsServerEvent::FRAME_RECIEVE to be called just after the Frame is obtained from the swoole server ( ... ?) and just before the Frame is sent via the swoole server (WebSocketServer::sendTo ?).
This way we can easly support websocket extensions like "permessage-deflate":

```php

/**
 * Class MessageSendRecieveListener
 * @Subscriber()
 */
class MessageSendRecieveListener implements EventSubscriberInterface
{

    public static function getSubscribedEvents(): array
    {
        return [
            WsServerEvent::MESSAGE_SEND     => 'handleSend',
            WsServerEvent::MESSAGE_RECIEVE  => 'handleRecieve',
            WsServerEvent::HANDSHAKE_BEFORE => 'handleHandshake'
        ];
    }

    /**
     * @param EventInterface $event
     */
    public function handleHandshake(EventInterface $event): void
    {
        $request = $event->getParam(0);
        $response = $event->getParam(1);

        if ($extensions = $request->header["sec-websocket-extensions"] ?? false) {
            $extensions = array_flip(array_map('trim', explode(";", $extensions)));
            if (isset($extensions['permessage-deflate'])) {
                $response->header("Sec-WebSocket-Extensions", "permessage-deflate");
                Session::mustGet()->set("permessage-deflate", true);
            }
        }
    }

    /**
     * @param EventInterface $event
     */
    public function handleRecieve(EventInterface $event): void
    {
        if(Session::mustGet()->get("permessage-deflate", false)) {
            $frame      = $event->getTarget();
            $ctx  = inflate_init(ZLIB_ENCODING_RAW);
            $frame->data = inflate_add($ctx, $frame->data, ZLIB_FINISH);
        }

    }

    /**
     * @param EventInterface $event
     */
    public function handleSend(EventInterface $event): void
    {
        if(Session::mustGet()->get("permessage-deflate", false)) {
            $frame      = $event->getTarget();
            $ctx        = deflate_init(ZLIB_ENCODING_RAW);
            $frame->data = deflate_add($ctx, $frame->data, ZLIB_FINISH);
        }

    }
}
```

**Comment by inhere** [comment]
ok, I'll think about it @goddanao

**Comment by inhere** [comment]
hi @goddanao Swoft v2.0.6 has been released.

more see https://itnext.io/php-microservice-framework-swoft-2-0-6-officially-released-602d0ace092c
